benchmarks/utils.py

```diff
@@ -0,0 +1,336 @@
+
+
+def _update_data_with_mask(data, struct, out=None):
+    """Update ``data`` with ``struct`` where ``struct`` is nonzero."""
+    # these branches are needed because np.where does not support
+    # an out= keyword argument
+    if out is None:
+        return np.where(struct, struct, data)
+    else:  # noqa: RET505
+        nz = struct != 0
+        out[nz] = struct[nz]
+        return out
+
+
+def _smallest_dtype(n: int) -> np.dtype:
+    """Find the smallest dtype that can hold n values."""
+    for dtype in [np.uint8, np.uint16, np.uint32, np.uint64]:
+        if np.iinfo(dtype).max >= n:
+            return dtype
+            break
+    else:
+        raise ValueError(f"{n=} is too large for any dtype.")
+
+
+@overload
+def labeled_particles(
+    shape: Sequence[int],
+    dtype: Optional[np.dtype] = None,
+    n: int = 144,
+    seed: Optional[int] = None,
+    return_density: Literal[False] = False,
+) -> np.ndarray: ...
+
+
+@overload
+def labeled_particles(
+    shape: Sequence[int],
+    dtype: Optional[np.dtype] = None,
+    n: int = 144,
+    seed: Optional[int] = None,
+    return_density: Literal[True] = True,
+) -> tuple[np.ndarray, np.ndarray, np.ndarray]: ...
+
+
+@lru_cache
+def labeled_particles(
+    shape: Sequence[int],
+    dtype: Optional[np.dtype] = None,
+    n: int = 144,
+    seed: Optional[int] = None,
+    return_density: bool = False,
+) -> Union[np.ndarray, tuple[np.ndarray, np.ndarray, np.ndarray]]:
+    """Generate labeled blobs of given shape and dtype.
+
+    Parameters
+    ----------
+    shape : Sequence[int]
+        Shape of the resulting array.
+    dtype : Optional[np.dtype]
+        Dtype of the resulting array.
+    n : int
+        Number of blobs to generate.
+    seed : Optional[int]
+        Seed for the random number generator.
+    return_density : bool
+        Whether to return the density array and center coordinates.
+    """
+    if dtype is None:
+        dtype = _smallest_dtype(n)
+    rng = np.random.default_rng(seed)
+    ndim = len(shape)
+    points = rng.integers(shape, size=(n, ndim))
+    # create values from 1 to max of number of points
+    values = np.linspace(1, n, n, dtype=dtype)
+    rng.shuffle(values)
+    # values = rng.integers(
+    #     np.iinfo(dtype).min + 1, np.iinfo(dtype).max, size=n, dtype=dtype
+    # )
+    sigma = int(max(shape) / (4.0 * n ** (1 / ndim)))
+    ball = _generate_ball(sigma, ndim)
+
+    labels = _structure_at_coordinates(
+        shape,
+        points,
+        ball,
+        multipliers=values,
+        reduce_fn=_update_data_with_mask,
+        dtype=dtype,
+    )
+
+    if return_density:
+        dens = _generate_density(sigma * 2, ndim)
+        densities = _structure_at_coordinates(shape, points, dens, reduce_fn=np.maximum, dtype=np.float32)
+
+        return labels, densities, points, values
+    else:  # noqa: RET505
+        return labels
+
+
+def run_benchmark_from_module(module: ModuleType, klass_name: str, method_name: str):
```

Comment:
explain purpose of this function (so that we can call the asv benchmarks manually and debug them, otherwise they are triggered by asv only)